ceph_deploy/gatherkeys.py

```diff
@@ -1,84 +1,254 @@
+        distro.conn,
+        arguments
+        )
+    if code != 0:
+        rlogger.error('"ceph auth get-or-create for keytype %s returned %s' % (keytype, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
+    keyring_name_local = keytype_path_to(args, keytype)
+    keyring_path_local = os.path.join(dest_dir, keyring_name_local)
+    with file(keyring_path_local, 'w') as f:
+        for line in out:
+            f.write(line + '\n')
+    return True
+
+
+def gatherkeys_with_mon(args, host, dest_dir):
+    """
+    Connect to mon and gather keys if mon is in quorum.
+    """
+    distro = hosts.get(host, username=args.username)
+    dir_keytype_mon = ceph_deploy.util.paths.mon.path(args.cluster, host)
+    path_keytype_mon = "%s/keyring" % (dir_keytype_mon)
+    mon_key = distro.conn.remote_module.get_file(path_keytype_mon)
+    if mon_key is None:
+        LOG.warning("No mon key found. Is '%s' a mon node" % (host))
+        return False
+    mon_name_local = keytype_path_to(args, "mon")
+    mon_path_local = os.path.join(dest_dir, mon_name_local)
+    with file(mon_path_local, 'w') as f:
+        f.write(mon_key)
+    rlogger = logging.getLogger(host)
+    path_asok = ceph_deploy.util.paths.mon.asok(args.cluster, host)
+    out, err, code = remoto.process.check(
+        distro.conn,
+            [
+                "/usr/bin/ceph",
+                "--connect-timeout=25",
+                "--cluster={cluster}".format(
+                    cluster=args.cluster),
+                "--admin-daemon={asok}".format(
+                    asok=path_asok),
+                "mon_status"
+            ]
+        )
+    if code != 0:
+        rlogger.error('"ceph mon_status %s" returned %s' % (host, code))
+        for line in err:
+            rlogger.debug(line)
+        return False
+    try:
+        mon_status = json.loads("".join(out))
+    except ValueError:
+        rlogger.error('"ceph mon_status %s" output was not json' % (host))
+        return False
+    mon_number = None
+    mon_map = mon_status.get('monmap')
+    if mon_map is None:
+        rlogger.error("could not find mon map for mons on '%s'" % (host))
+        return False
+    mon_quorum = mon_status.get('quorum')
+    if mon_quorum is None:
+        rlogger.error("could not find quorum for mons on '%s'" % (host))
+        return False
+    mon_map_mons = mon_map.get('mons')
+    if mon_map_mons is None:
+        rlogger.error("could not find mons in monmap on '%s'" % (host))
+        return False
+    for mon in mon_map_mons:
+        if mon.get('name') == host:
+           mon_number = mon.get('rank')
+           break
+    if mon_number is None:
+        rlogger.error("could not find '%s' in monmap" % (host))
+        return False
+    if not mon_number in mon_quorum:
+        rlogger.error("Not yet quorum for '%s'" % (host))
+        return False
+    got_all_keys = True
+    for keytype in ["admin", "mds", "osd", "rgw"]:
```

Comment:
this loops over 4 different key types and sets `got_all_keys` at every loop. What happens if the `osd` keytype is not found, the code sets `got_all_keys` to `False` but `rgw` keys are found so it sets it back to `True`? It would return `got_all_keys` as `True` even when `osd` keys are not found.

Is that intended? Do we care what `gatherkeys_missing` is actually collecting? This is a bit of bookkeeping that it isn't used (or used improperly)